Reorganise C parameter creation

owen-jones-diffblue opened the pull request:
Make it match the java code more closely to prepare for (a) one day
merging them and (b) dealing with pointer parameters.

owen-jones-diffblue commented:
Note that this patch makes one test fail: cbmc/unsigned__int128. I paste below a description of the problem that I put on slack.

tautschnig commented:
@owen-jones-diffblue Brilliant, thank you for debugging and *actually* fixing the test!!!

Would you mind, however, placing this in a separate PR? I believe this should be merged ASAP (and it's really unrelated to your additional work).

owen-jones-diffblue commented:
Yes, I'd just come to the same conclusion

tautschnig commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -40,66 +41,41 @@ Function: build_function_environment
-
-    result[i]=symbol_expr;
-    i++;
+    const code_typet::parametert &p = parameters[param_number];
```
Nit picking: no whitespace around `=`

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -40,66 +41,41 @@ Function: build_function_environment
+      init_code,
+      allow_null,
+      symbol_table,
+      p.source_location(),
```
This function has a large number of parameters, and seemingly at least 2 can be derived from the first one. Can this be simplified?

owen-jones-diffblue replied:
I was trying to keep it similar to object_factory() in src/java_bytecode/java_object_factory.cpp so that they can be easily combined in future. If it is preferable then I can instead simplify it and reduce the number of arguments.

tautschnig replied:
This makes a lot of sense. I'm wondering whether the Java code should be refactored instead?

owen-jones-diffblue replied:
I looked into refactoring this aspect of the java code. object_factory() is used in two places in java_entry_point(): once in a loop over all the parameters of the function we are looking at, and once in a loop over all symbols in the symbol table which are static plus some other conditions, or which are string literals. Without knowing the code better it's not clear to me that I can reduce the number of parameters in a way which works in both places.

tautschnig replied:
Ok, thanks for checking!

tautschnig commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -40,66 +41,41 @@ Function: build_function_environment
+
+    main_arguments[param_number]=c_nondet_symbol_factory(
+      p,
+      param_number,
```
Why is the position of the parameter a required piece of information?

owen-jones-diffblue replied:
It is only used to construct a fallback base name which is used if the parameter doesn't have one. That is what the code did before I refactored it. Other options include passing in the fallback base name or trying to establish that the parameter will always have a base name.

tautschnig replied:
Ok; I'm inclined to suggest that the name should be passed in (effectively as constructed in that function) as the problem can only arise at the top level, recursive calls would not suffer from that problem.

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -40,66 +41,41 @@ Function: build_function_environment
+    input_recorder.record_input(
+      p_symbol.base_name,
+      main_arguments[param_number]);
+    input_recorder.declare_recorded_inputs(init_code);
```
The input recorder is passed as an argument to `c_nondet_symbol_factory`, but then there still is a need for an extra manual step. I think this should be simplified.

owen-jones-diffblue replied:
You are right. I have moved this entirely within c_nondet_symbol_factory().

tautschnig commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -505,8 +481,11 @@ bool ansi_c_entry_point(
     // produce nondet arguments
-    call_main.arguments()=
-      build_function_environment(parameters, init_code, symbol_table);
+    call_main.arguments()=build_function_environment(
```
Nit pick: Preferably wrap lines after the `=` (applies elsewhere in your patch too).

owen-jones-diffblue replied:
I have added an argument (message_handler, to make it more like the java code), so it no longer fits into one line if I wrap after the `=`.

tautschnig replied:
Ah, apologies for being unclear: *also* wrap after the `=` (and then indent the arguments by two additional spaces).

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+class symbol_factoryt:public messaget
+{
+  code_blockt &init_code;
+  std::set<irep_idt> recursion_set;
```
Very likely this should be a `std::unordered_set`

owen-jones-diffblue replied:
This variable isn't used, so I've deleted it. It's used in the java code to cope with structs, which the C code doesn't do yet.

owen-jones-diffblue replied:
I've updated this in the java code

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+      init_code.move_to_operands(init_done_label);
+    }
+  }
+  else
```
What about structs, arrays of pointers?

owen-jones-diffblue replied:
This patch is only intended to make pointers work. Structs and arrays need to be done, but I think it is best left to another patch.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+      auto assign_returns_null=
+          code_assignt(returns_null, c_get_nondet_bool(returns_null_sym.type));
+      assign_returns_null.add_source_location()=loc;
+      init_code.move_to_operands(assign_returns_null);
```
Why not just use `side_effect_expr_nondett`?

owen-jones-diffblue replied:
c_get_nondet_bool() does use side_effect_expr_nondett. I've copied from this function, and this use of it, from the java code.

tautschnig replied:
Ah, I'm just wondering why a new symbol needs to be introduced, instead of just using the nondet directly?

owen-jones-diffblue replied:
Just my inexperience. I've changed it to condition on the return value c_get_nondet_bool() directly.

tautschnig replied:
You may indeed want to submit improvements to the Java input generator!

owen-jones-diffblue replied:
I will submit a number of changes. Should I make it part of this PR or separate?

tautschnig replied:
I think you can safely place them in here, and then we can also get the original authors of the Java one as reviewers for the overall code.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+    static size_t synthetic_constructor_count=0;
+
+    if(!assume_non_null)
+    {
```
It would be nice to have a comment that sketches the generated code as pseudo-code.

owen-jones-diffblue replied:
Good idea. I've tried to do this.

tautschnig replied:
You haven't pushed the new version yet, have you? I was hoping for comments along the lines of [goto_convertt::convert_for](https://github.com/diffblue/cbmc/blob/master/src/goto-programs/goto_convert.cpp#L1245-L1254)

owen-jones-diffblue replied:
No, I haven't pushed yet. I'm writing comments here as I address things and then I'll push at the end. I hope that's okay.

tautschnig replied:
Of course - I just wasn't entirely sure how to judge whether I'll find it readable ;-)

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+    code_labelt set_null_label;
+    code_labelt init_done_label;
+
+    static size_t synthetic_constructor_count=0;
```
Locality: keep the declaration within the scope that actually uses it.

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+
+      code_ifthenelset null_check;
+      exprt null_return=
+        zero_initializer(returns_null_sym.type, loc, ns, message_handler);
```
You should be able to get away with `from_integer(0, returns_null_sym.type)`, at which point you might not even need a `message_handlert` anymore.

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+    type,
+    prefix);
+
+  code_declt decl;
```
`code declt decl(tmp_symbol.symbol_expr());`

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+  symbol_exprt symbol_expr(identifier, type);
+
+  code_declt decl;
+  decl.symbol()=symbol_expr;
```
`code_declt decl(symbol_expr);`

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,395 @@
+      message_handler,
+      input_recorder);
+    state.gen_nondet_init(symbol_expr, loc);
+  }
```
Why just those to type classes?

owen-jones-diffblue replied:
I copied the java code. The pointer case seems to cover arrays and structs as well, though this patch doesn't properly deal with them.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.h:
```diff
@@ -0,0 +1,64 @@
+  symbol_tablet &symbol_table,
+  const source_locationt &loc,
+  const typet &type,
+  const std::string &prefix="tmp");
```
The above appear to be helper functions that ought to be just marked static and not be exported via the header file either.

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -40,66 +41,34 @@ Function: build_function_environment
-    i++;
+    const code_typet::parametert &p=parameters[param_number];
+
+    bool allow_null=true;
```
Is there a reason that this is placed in a local variable rather than passing `true` directly?

owen-jones-diffblue replied:
No, I guess it'll always be true here. I was thinking that when we add arrays and structs we'd need to have some logic to choose it, but I guess not at this level.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,403 @@
+    if(!assume_non_null)
+    {
+      // Will generate the following code:
+      //   IF NONDET(_Bool) != FALSE THEN GOTO <set_null_label_str>
```
Thanks a lot for the detailed comments. Could you please add the overall code in one place to understand what's intended to happen here at high level?

owen-jones-diffblue replied:
I'll move them all into one place, rather than have them in one place and spread out.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,403 @@
+    new_symbol.base_name=base_name;
+    new_symbol.type=type;
+
+    symbol_table.move(new_symbol);
```
Use symbol_table.add and make sure this symbol is actually new (add has a return value).

owen-jones-diffblue replied:
Done. I also wrote the function comments for symbol_table.add() and symbol_table.move(), which I'll put in a separate PR.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,404 @@
+      //   <code generated by gen_nondet_init()>
+      //   GOTO <init_done_label_str>
+      //   <set_null_label_str>: <expr> = ((<pointer_type>)NULL);
+      //   <init_done_label_str>: SKIP
```
Thanks a lot for adding those comments! It seems the code here really mixes up two concepts (I understand that you've just copied this from the Java version): the code is being generated as if in the GOTO intermediate level, but it's being done using high-level language constructs. There does not seem to be any need for labels or gotos here. Just generate a `code_ifthenelse` with a `then_case` and an `else_case`. This avoids the static variable and synthetic labels, and will thus make the code much more concise.

owen-jones-diffblue replied:
I've done this. I had to make the auxiliary variables static, otherwise they were going out of scope at the end of the else clause, being declared dead, and CBMC didn't produce the right test suite.

owen-jones-diffblue commented:
I haven't made any changes to the java code since the last round - I'll do it once the C code has been reviewed.

I included a commit to turn off the failing test that is fixed in #771 so that the CI can produce the correct results. This commit should be removed before this PR is merged.

Using code_ifthenelset properly makes things a lot simpler. Unfortunately it introduced issues of scope, so I've had to make all the intermediate variables have is_static_lifetime true so they don't go out of scope at the end of the else clause. This means that they aren't declared in the output of show-goto-functions, which I feel is a bit less clear, but seems to be the desired behaviour for is_static_lifetime (=global?) variables. Since the declarations are not doing anything I've removed them. Java already uses is_static_lifetime=true intermediate variables, I believe. One alternative would be to collect all the variable declarations in a separate code block and output them before all the other code for that parameter, in a similar way that we collect the INPUT() lines to the end.

I made a helper function, concatenate_code_blocks(), which I feel should be a method on code_blockt.

tautschnig commented:
> I haven't made any changes to the java code since the last round - I'll do it once the C code has been reviewed.

Cool, I'll take a look.

> I included a commit to turn off the failing test that is fixed in #771 so that the CI can produce the correct results. This commit should be removed before this PR is merged.

I've taken another look at #771, and merged it to avoid further overhead. You might thus remove your commit (after rebasing).

> Using code_ifthenelset properly makes things a lot simpler. Unfortunately it introduced issues of scope, so I've had to make all the intermediate variables have is_static_lifetime true so they don't go out of scope at the end of the else clause. This means that they aren't declared in the output of show-goto-functions, which I feel is a bit less clear, but seems to be the desired behaviour for is_static_lifetime (=global?) variables. Since the declarations are not doing anything I've removed them. Java already uses is_static_lifetime=true intermediate variables, I believe. One alternative would be to collect all the variable declarations in a separate code block and output them before all the other code for that parameter, in a similar way that we collect the INPUT() lines to the end.

Ok, I'll take a look. I'm pretty sure that the declarations should just be placed ahead of any if/else anyway.

> I made a helper function, concatenate_code_blocks(), which I feel should be a method on code_blockt.

I fully agree.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,419 @@
+
+\*******************************************************************/
+
+static symbolt &c_new_tmp_symbol(
```
I think you can return it as a `const symbolt &`

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,419 @@
+  {
+    target_block.add(to_code(operand));
+  }
+}
```
As suggested in your comment, this should really go in util/std_code.h, as a member function of `code_blockt`. It would then be worthwhile using `reserve` to do the re-allocation of memory just once.

owen-jones-diffblue replied:
Done

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,419 @@
+  //   INPUT("tmp$<temporary_counter>", tmp$<temporary_counter>);
+
+  symbolt &aux_symbol=c_new_tmp_symbol(symbol_table, loc, allocate_type);
+  aux_symbol.is_static_lifetime=static_lifetime;
```
Maybe static lifetime should actually be a parameter to `c_new_tmp_symbol`? setting it in there once and that re-setting it here seems questionable.

owen-jones-diffblue replied:
Oops, that was a mistake. Well, hopefully I'll stop using static lifetime tmp symbols, but if I do continue using them I'll make it an option, and definitely only do it in one place.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,419 @@
+    else
+    {
+      // Add the following code to init_code:
+      //           IF !(NONDET(_Bool) == FALSE) THEN GOTO <label1>
```
This comment isn't exactly what the code now does, please update.

owen-jones-diffblue replied:
I'm not sure exactly what you mean. Do you mean that changing the condition of the if-then-else clause means that this line changes? I ended up having to back out that change because it wouldn't run with just --cover branch, without --show-goto-functions.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,419 @@
+      const exprt null_return=from_integer(0, c_bool_typet(1));
+
+      code_ifthenelset null_check;
+      null_check.cond()=equal_exprt(nondet_bool, null_return);
```
This equality test on nondet is redundant - it's sufficient to just do `null_check.cond()=c_get_nondet_bool(c_bool_typet(1));`

owen-jones-diffblue replied:
Done. I've been thinking that the code that was being generated here was very clunky, but I didn't see the obvious fix.

owen-jones-diffblue replied:
Actually that generated an error when running without --show-goto-functions. I ran on pointer-function-parameters/main.c with the flags --function fun --cover branch and I got this output:
```
Passing problem to propositional reduction
converting SSA
######### lhs: symbol
  * type: bool
  * identifier: goto_symex::\guard#1
  * expression: symbol
      * type: bool
      * identifier: goto_symex::\guard
  * L2: 1
  * L1_object_identifier: goto_symex::\guard
  * #SSA_symbol: 1
######### rhs: typecast
  * type: c_bool
      * width: 1
  0: nondet_symbol
      * type: bool
      * identifier: symex::nondet0
      * #source_location: nil
equality without matching types
/home/owen/workspace/cbmc/src/cbmc/cbmc exited with code 10
```

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,419 @@
+    // pointer to a pointer temporary variables are created in else clauses,
+    // and thus go out of scope before the function is called.
+    exprt allocated=
+      allocate_object(non_null_inst, expr, subtype, true);
```
If you make this call return two code blocks -- a declaration block and an assignment block, then you should be able to get away without static-lifetime objects. This would be highly preferable for the concurrency encoding.

owen-jones-diffblue replied:
It's a little bit more complicated than that. Because when this function recursively calls itself, even if we put the variable declaration before the if-else statement it can end up inside another else clause, and thus going out of scope before we call the entry function. I can however do it by collecting all the variable declarations separately from the other code and then putting them in the right order at the very end, a bit like what we're currently doing with input_recorder.

owen-jones-diffblue commented:
I've updated the PR addressing all comments. I realised I could deal with putting all the variable declarations first and putting all the INPUT statements last at the same time by keeping a list of pointers to the symbols that had been created and generating all the code at the very end. That removes all the code for input_recordert, which makes the code a lot shorter.

As before, I haven't updated the java code, but I will do so once it's clear that the C code is taking the right approach.

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -0,0 +1,330 @@
+
+      code_ifthenelset null_check;
+      null_check.cond()=
+        equal_exprt(c_get_nondet_bool(c_bool_typet(1)), false_expr);
```
To address the equality-without-matching-types issue you should probably just do
```
null_check.cond()=side_effect_expr_nondett(bool_typet());
```
The issue is the `c_bool_typet` placed in here. Apologies for not spotting that any earlier.

owen-jones-diffblue replied:
Done

owen-jones-diffblue commented:
Yes, I definitely want to clean up the java, especially the simpler use of if_then_elset.

owen-jones-diffblue commented:
I created #853 to do the final refactoring I'd intended.
